seperation of network call in KubernetesReleaseVersion

Klaven opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
Made network calls package local so unit tests would be able to mock them out. I did not put them behind a full interface due to it adding unnecessary complexity for something we are not even sure should ever be supported (fetching the version from someplace other then a URL.)

**Which issue(s) this PR fixes**:
Fixes #https://github.com/kubernetes/kubeadm/issues/902

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

Commit:
seperation of network calls when getting version

updated the network calls to be package local so tests could pass their
own implementation. A public interface was not provided as it would not
be likely this would ever be needed or wanted.

Klaven commented:
no. don't test! :(

rosti commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -61,7 +100,7 @@ var (
 //  latest-1    (latest release in 1.x, including alpha/beta)
 //  latest-1.0  (and similarly 1.1, 1.2, 1.3, ...)
-func KubernetesReleaseVersion(version string) (string, error) {
+func KubernetesReleaseVersion(version string, fetcher URLFetcher) (string, error) {
```
The whole idea of the interface seems like an overkill to me. You need just a single function. It's unlikely (at least for me) that `KubernetesReleaseVersion` will require something different for mocking. Also, ideally we don't want users of `KubernetesReleaseVersion` to construct and pass custom fetchers.

What I propose is to move the entire body of `KubernetesReleaseVersion` to a package local implementation function that takes fetcher func as an additional parameter. Then simply implement `KubernetesReleaseVersion` as a one liner that calls the internal implementation with `fetchFromURL`.

Thus the test cases can now use the internal implementation with mocked URL fetcher and provide tests for the edge cases.
Here is something similar that I did a while ago:
https://github.com/kubernetes/kubernetes/blob/46a80259f681736b4669efa49304bcb4218bab34/cmd/kubeadm/app/util/runtime/runtime.go#L235

neolit123 commented:
cc

Klaven commented:
/test pull-kubernetes-e2e-gce

rosti commented on cmd/kubeadm/app/util/version.go:
```diff
@@ -62,6 +62,10 @@ var (
+	return releaseVersion(version, fetchFromURL)
+}
+
+func releaseVersion(version string, fetcher func(string, time.Duration) (string, error)) (string, error) {
```
I'd rather want the name to contain the word "kubernetes" or "k8s" somewhere. This is not a local func for the file, but local for the whole util package, hence it needs more context.

rosti reviewed (commented):
Thanks @Klaven !
That's what I expected. However, let's have it along with some tests, that actually make use of the new way of mocking fetching a version over the internet.

Klaven commented:
/test pull-kubernetes-kubemark-e2e-gce-big

Klaven commented:
/test pull-kubernetes-kubemark-e2e-gce-big

rosti commented:
/priority important-longterm
/assign

rosti reviewed (commented):
Thanks @Klaven !
As a few ideas for additional test cases, you can try to use the new func in place of the full blown one in `version_test.go`. For example `TestVersionFromNetwork` can be done without running mock HTTP server, `TestValidVersion` should never go on the network, and so on.

rosti reviewed (commented):
On a good track here. You can also try to modify the `TestValidVersion`, `TestInvalidVersion` and `TestEmptyVersion` to fail if they try to access the internet.

Commit:
updated tests to prevent false positive

one test also proved it did not call the internet
but this was not fool proof as it did not return a string
and thus could be called with something expecting to fail.

Commit:
update to remove unused test

commit will be squashed before merge

Klaven commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -121,22 +121,20 @@ func TestVersionFromNetwork(t *testing.T) {
 		res, found := cases[key]
 		if found {
-			http.Error(w, res.Content, res.Status)
+			return res.Content, nil
```
I do not think Status needs taken into account and was really just used for the web server part. So I removed it. this also allowed the removal of the `net/http` import as you stated.

rosti replied:
Still, the `unknown` test case above, should result in an error to be returned by the fetcher func.

Klaven commented:
not sure what is going on with bazel. I rebased with master, re-ran `./hack/update-bazel.sh`... am I missing something? I have never seen it be flaky but I cant seem to see it's logs.

rosti commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -50,7 +50,10 @@ func TestValidVersion(t *testing.T) {
 		t.Run(s, func(t *testing.T) {
-			ver, err := KubernetesReleaseVersion(s)
+			fileFetcher := func(url string, timeout time.Duration) (string, error) {
+				return "", errors.New("should not make internet call")
```
You can add the `url` to the error message.

rosti commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -72,7 +75,10 @@ func TestInvalidVersion(t *testing.T) {
 	for _, s := range invalidVersions {
 		t.Run(s, func(t *testing.T) {
-			ver, err := KubernetesReleaseVersion(s)
+			fileFetcher := func(url string, timeout time.Duration) (string, error) {
```
You can reuse the same func from `TestValidVersion`. Probably extract it in the file and give it a name.

rosti commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -92,7 +98,10 @@ func TestValidConvenientForUserVersion(t *testing.T) {
 	for _, s := range validVersions {
 		t.Run(s, func(t *testing.T) {
-			ver, err := KubernetesReleaseVersion(s)
+			fileFetcher := func(url string, timeout time.Duration) (string, error) {
```
Again, reuse?

rosti commented on cmd/kubeadm/app/util/version_test.go:
```diff
@@ -272,7 +277,15 @@ func TestCIBuildVersion(t *testing.T) {
+				if tc.valid {
+					return tc.expected, nil
+				}
+				return "Unknown test case key!", fmt.Errorf("unknown test case key")
```
s/Errorf/Error

Commit:
updates based off reviews

Commit:
added comment

rosti reviewed (approved):
Thanks @Klaven !
/approve
/lgtm
